### PR TITLE
Switch to ValidatedNel / Validated instead of Either

### PR DESCRIPTION
**Pull request by sihil**
This makes much of the code a bit cleaner and easier to reason about and in come cases means that we can return multiple errors. As a bonus some of the Json reads code becomes a little easier to reason about as well.

This adds an API endpoint to which YAML can be sent for validation and some **very** rudimentary web pages that let you do the same (your eyes will hurt!)

Would be interested in doing a code review of all of this in person with anyone who has used cats before.

### CODE REVIEW

File: riff-raff/app/controllers/api.scala [hunk 1]
```diff
@@ -347,4 +348,52 @@ class Api(deployments: Deployments)(implicit val messagesApi: MessagesApi, val w
...
+                  "dependencies" -> deployment.dependencies,
+                  "parameters" -> deployment.parameters
+                )
+              }
+            )
+          )
+        case Invalid(errors) =>
+          Json.obj(
+            "response" -> Json.obj(
+              "status" -> "ok",
```
Comment (adamnfish) [inline riff-raff/app/controllers/api.scala]: Is the status really "ok" here?

File: riff-raff/app/views/validation/validationErrors.scala.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
+@import cats.data.NonEmptyList
+@import magenta.input.ConfigError
+@(implicit request: Security.AuthenticatedRequest[AnyContent, com.gu.googleauth.UserIdentity], errors: NonEmptyList[ConfigError])
+
+@main("Validation errors", request) {
+    <div class="clearfix"><p>&nbsp;</p></div>
```
Comment (adamnfish) [inline riff-raff/app/views/validation/validationErrors.scala.html]: :-0

**Comment by adamnfish** [comment]
Woah, `NEL`ly

File: riff-raff/app/controllers/api.scala (hunk 1 above)
Comment (sihil) [inline riff-raff/app/controllers/api.scala]: Yes. I *think* so. It has successfully given you a validation result. That result might be a failure, but it is still a 200 response which in the API world is OK. See the case above where you don't even give it valid data to validate in which case you do get an error.

File: magenta-lib/src/main/scala/magenta/input/resolver/TaskResolver.scala [hunk 3]
```diff
@@ -8,15 +9,15 @@ import magenta.{App, DeployParameters, DeployTarget, DeploymentPackage, Deployme
 
 object TaskResolver {
   def resolve(deployment: Deployment, deploymentResources: DeploymentResources, parameters: DeployParameters,
-    deploymentTypes: Seq[DeploymentType], artifact: S3Artifact): Either[ConfigError, DeploymentTasks] = {
+    deploymentTypes: Seq[DeploymentType], artifact: S3Artifact): ValidatedNel[ConfigError, DeploymentTasks] = {
     val deploymentPackage = createDeploymentPackage(deployment, artifact)
-    val deploymentTypeEither = deploymentTypes.find(_.name == deployment.`type`).
-      toRight(ConfigError(deployment.name, s"Deployment type ${deployment.`type`} not found"))
+    val validatedDeploymentType = Validated.fromOption(deploymentTypes.find(_.name == deployment.`type`),
+      NEL.of(ConfigError(deployment.name, s"Deployment type ${deployment.`type`} not found")))
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/resolver/TaskResolver.scala]: Yep - that was before I started aliasing `NonEmptyList => NEL` in all the imports. Am thinking about a global type alias.

File: magenta-lib/src/test/scala/magenta/fixtures/ValidatedValues.scala [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import org.scalatest.exceptions.TestFailedException
+
+/**
+  * This trait is inspired strongly by [[org.scalatest.EitherValues]] as a way of simply asserting that you expect
+  * either a Valid or Invalid value from a Validated type.
+  *
+  * Two functions are pimped onto Validated for the purposes of testing and if the value is not of the expected type
+  * then the test will fail detailing what the unexpected value was.
+  */
+trait ValidatedValues {
```
Comment (sihil) [inline magenta-lib/src/test/scala/magenta/fixtures/ValidatedValues.scala]: I've left it as I can't think of an alternative.

File: magenta-lib/src/main/scala/magenta/input/models.scala [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
 
+import cats.data.NonEmptyList
 import play.api.libs.json._
 
 case class ConfigError(context: String, message: String)
-case class ConfigErrors(errors: List[ConfigError]) {
-  def :::(other: ConfigErrors) = ConfigErrors(other.errors ::: errors)
-  def ::(other: ConfigError) = ConfigErrors(other :: errors)
+object ConfigError {
+  def nel(context: String, message: String) = NonEmptyList.of(ConfigError(context, message))
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/models.scala]: No - have removed.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala [hunk 6]
```diff
@@ -1,38 +1,40 @@
...
-  def resolve(config: RiffRaffDeployConfig): List[Either[ConfigError, Deployment]] = {
+  def resolve(config: RiffRaffDeployConfig): ValidatedNel[ConfigError, List[Deployment]] = {
     config.deployments.map { case (label, rawDeployment) =>
-      for {
-        templated <- applyTemplates(label, rawDeployment, config.templates).right
-        deployment <- resolveDeployment(label, templated, config.stacks, config.regions).right
-        _ <- validateDependencies(label, deployment, config.deployments).right
-      } yield deployment
-    }
+      applyTemplates(label, rawDeployment, config.templates).andThen { templated =>
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: FIXED!

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentTypeResolver.scala [hunk 7]
```diff
@@ -1,17 +1,19 @@
...
-  def validateDeploymentType(deployment: Deployment, availableTypes: Seq[DeploymentType]): Either[ConfigError, Deployment] = {
-    for {
+  def validateDeploymentType(deployment: Deployment, availableTypes: Seq[DeploymentType]): ValidatedNel[ConfigError, Deployment] = {
+    val result = for {
       deploymentType <- availableTypes.find(_.name == deployment.`type`)
         .toRight(ConfigError(deployment.name, s"Unknown type ${deployment.`type`}")).right
       deploymentWithActions <- resolveDeploymentActions(deployment, deploymentType).right
       deployment <- verifyDeploymentParameters(deploymentWithActions, deploymentType).right
     } yield deployment
+    Validated.fromEither(result).leftMap(NonEmptyList.of(_))
```
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentTypeResolver.scala]: Sadly not.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala [hunk 8]
```diff
@@ -1,38 +1,41 @@
...
+import cats.syntax.cartesian._
+import cats.syntax.traverse._
+import cats.instances.list._
 import magenta.input.{ConfigError, Deployment, DeploymentOrTemplate, RiffRaffDeployConfig}
 
 object DeploymentResolver {
-  def resolve(config: RiffRaffDeployConfig): List[Either[ConfigError, Deployment]] = {
-    config.deployments.map { case (label, rawDeployment) =>
+  def resolve(config: RiffRaffDeployConfig): ValidatedNel[ConfigError, List[Deployment]] = {
+    config.deployments.traverseU[ValidatedNel[ConfigError, Deployment]] { case (label, rawDeployment) =>
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: https://github.com/scala/scala/pull/5102 means that the weird `U` bit will go away in 2.12.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala [hunk 9]
```diff
@@ -1,38 +1,41 @@
...
     * deployment attributes with any globally defined properties.
     */
-  private[input] def resolveDeployment(label: String, templated: DeploymentOrTemplate, globalStacks: Option[List[String]], globalRegions: Option[List[String]]): Either[ConfigError, Deployment] = {
-    for {
-      deploymentType <- templated.`type`.toRight(ConfigError(label, "No type field provided")).right
-      stacks <- templated.stacks.orElse(globalStacks).toRight(ConfigError(label, "No stacks provided")).right
-      regions <- templated.regions.orElse(globalRegions).toRight(ConfigError(label, "No regions provided")).right
-    } yield {
+  private[input] def resolveDeployment(label: String, templated: DeploymentOrTemplate, globalStacks: Option[List[String]], globalRegions: Option[List[String]]): ValidatedNel[ConfigError, Deployment] = {
+    (Validated.fromOption(templated.`type`, NEL.of(ConfigError(label, "No type field provided"))) |@|
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: I think a custom error type with it's own `Semigroup` might avoid the littering of `NEL.of`.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala [hunk 10]
```diff
@@ -1,38 +1,41 @@
...
+      Validated.fromOption(templated.stacks.orElse(globalStacks), NEL.of(ConfigError(label, "No stacks provided"))) |@|
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: It tend to find `mapN` easier to talk about than `|@|`, but that's definitely personal preference.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala [hunk 11]
```diff
@@ -1,38 +1,41 @@
...
-        stacks            = stacks,
-        regions           = regions,
-        actions           = templated.actions,
-        app               = templated.app.getOrElse(label),
-        contentDirectory  = templated.contentDirectory.getOrElse(label),
-        dependencies      = templated.dependencies.getOrElse(Nil),
-        parameters        = templated.parameters.getOrElse(Map.empty)
+        name = label,
+        `type` = deploymentType,
+        stacks = NEL.fromListUnsafe(stacks),
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: Couldn't this `Unsafe` be avoided by lifting into the NEL within the `Validated` clause? 

e.g.

```scala
Validated.fromOption(templated.stacks.orElse(globalStacks).flatMap(NEL.fromList), NEL.of(ConfigError(label, "No stacks provided")))
```

File: magenta-lib/src/main/scala/magenta/input/resolver/package.scala [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package magenta.input
+
+import cats.data.Validated
+
+package object resolver {
+  implicit class RichValidated[E, A](validated: Validated[E, A]) {
+    def flatMap[EE >: E, B](f: A => Validated[EE, B]): Validated[EE, B] = validated.andThen(f)
```
Comment (philwills) [inline magenta-lib/src/main/scala/magenta/input/resolver/package.scala]: I think there are some quite good reasons for not calling this `flatMap`

Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/resolver/package.scala]: The reason it is here is to make for comprehensions work and eliminate the uses of andThen across the codebase.

File: magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala (hunk 9 above)
Comment (sihil) [inline magenta-lib/src/main/scala/magenta/input/resolver/DeploymentResolver.scala]: Let me give that a bash.

File: riff-raff/app/views/validation/validationErrors.scala.html (hunk 2 above)
Comment (philwills) [inline riff-raff/app/views/validation/validationErrors.scala.html]: I think you really need a spacer.gif to get into the flavour.

**Comment by sihil** [comment]
OK - I've made some more changes to move from `ValidatedNel` to `Validated` with a custom error class. Cleans up quite a lot of the boilerplate as @philwills guessed. I also found a couple of places I hadn't entirely converted to `Validated`.

**Pull request merged**